### PR TITLE
adding dialyzer and addressing some of the issues

### PR DESCRIPTION
**Pull request by InoMurko**
Issue/Task Number:
/
# Overview

Adding dialyzer to eWallet for static code analysis.

# Changes
Dialyzer issues addressed (I hope all are in the list):
```
apps/admin_api/lib/admin_api/invites/inviter.ex:22: Expression produces a value of type #{'__struct__':='Elixir.Bamboo.Email', 'assigns':=#{atom()=>_}, 'bcc':=binary() | {binary(),binary()}, 'cc':=binary() | {binary(),binary()}, 'from':=binary() | {binary(),binary()}, 'headers':=#{binary()=>binary()}, 'html_body':='nil' | binary(), 'private':=#{atom()=>_}, 'subject':='nil' | binary(), 'text_body':='nil' | binary(), 'to':=binary() | {binary(),binary()}}, but this value is unmatched

apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex:32: Guard test 'or'('false','false') can never succeed

apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex:32: Guard test 'or'('false','false') can never succeed
apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex:33: The pattern _account@1 = #{'__struct__':='Elixir.EWalletDB.Account'} can never match the type 'nil' | {'error','account_id_not_found'}
apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex:73: Function assign_or_invite/4 will never be called
apps/admin_api/lib/admin_api/v1/controllers/account_membership_controller.ex:103: The pattern _account@1 = #{'__struct__':='Elixir.EWalletDB.Account'} can never match the type 'nil' | {'error','account_id_not_found'}

apps/admin_api/lib/admin_api/v1/controllers/minted_token_controller.ex:88: The pattern _minted_token@1 = #{'__struct__':='Elixir.EWalletDB.MintedToken'} can never match the type 'nil'
apps/admin_api/lib/admin_api/v1/controllers/minted_token_controller.ex:109: The pattern {'error', _code@1, _description@1} can never match the type {'error',_} | {'error',_,_,_}
apps/admin_api/lib/admin_api/v1/controllers/minted_token_controller.ex:135: The pattern <_minted_token@1 = #{'__struct__':='Elixir.EWalletDB.MintedToken'}, _conn@1> can never match the type <'nil',_>

apps/admin_api/lib/admin_api/v1/controllers/reset_password_controller.ex:17: Expression produces a value of type #{'__struct__':='Elixir.Bamboo.Email', 'assigns':=#{atom()=>_}, 'bcc':=binary() | {binary(),binary()}, 'cc':=binary() | {binary(),binary()}, 'from':=binary() | {binary(),binary()}, 'headers':=#{binary()=>binary()}, 'html_body':='nil' | binary(), 'private':=#{atom()=>_}, 'subject':='nil' | binary(), 'text_body':='nil' | binary(), 'to':=binary() | {binary(),binary()}}, but this value is unmatched
apps/admin_api/lib/admin_api/v1/controllers/reset_password_controller.ex:37: Expression produces a value of type atom() | #{'uuid':=_, _=>_}, but this value is unmatched

apps/admin_api/lib/admin_api/v1/controllers/user_controller.ex:58: The pattern <{'error', _changeset@1}, _conn@1> can never match the type <'nil',_>

apps/local_ledger/lib/local_ledger/cached_balance.ex:13: Invalid type specification for function 'Elixir.LocalLedger.CachedBalance':cache_all/0. The success typing is () -> 'ok'

apps/ewallet_db/lib/ewallet_db/validator.ex:104: The pattern {'error', _} can never match the type {'ok',binary()} | {'error','too_short',[{'min_length',8},...]}


apps/ewallet_db/lib/ewallet_db/minted_token.ex:150: The call 'Elixir.EWalletDB.MintedToken':get_by([{'id',_},...],opts@1::any()) breaks the contract (fields::map(),opts::elixir:keyword()) -> #{'__struct__':='Elixir.EWalletDB.MintedToken','__meta__':=term(),'account':=term(),'account_uuid':=term(),'description':=term(),'encrypted_metadata':=term(),'encryption_version':=term(),'html_entity':=term(),'id':=term(),'inserted_at':=term(),'iso_code':=term(),'iso_numeric':=term(),'locked':=term(),'metadata':=term(),'name':=term(),'short_symbol':=term(),'smallest_denomination':=term(),'subunit':=term(),'subunit_to_unit':=term(),'symbol':=term(),'symbol_first':=term(),'updated_at':=term(),'uuid':=term()} | 'nil'

apps/ewallet_db/lib/ewallet_db/api_key.ex:120: The pattern <'nil', _> can never match the type <bitstring(),atom()>

apps/ewallet_db/lib/ewallet_db/account.ex:193: The call 'Elixir.EWalletDB.Account':get_by([{'id',binary()},...],opts@1::any()) breaks the contract (fields::map(),opts::elixir:keyword()) -> #{'__struct__':='Elixir.EWalletDB.Account','__meta__':=term(),'api_keys':=term(),'avatar':=term(),'description':=term(),'encrypted_metadata':=term(),'encryption_version':=term(),'id':=term(),'inserted_at':=term(),'keys':=term(),'memberships':=term(),'metadata':=term(),'minted_tokens':=term(),'name':=term(),'parent':=term(),'parent_uuid':=term(),'relative_depth':=term(),'updated_at':=term(),'uuid':=term(),'wallets':=term()}
```

# Usage

mix dialyzer or mix help dialyzer 

There are quite a few others left but decided to do this incrementally. Smaller PR is easier to address then a large one and I can avoid merge conflicts.

### CODE REVIEW

**Comment by sirn** [comment]
Thank you for your contribution! CI is currently failing due to:

```
** (Mix) mix format failed due to --check-formatted.
The following files were not formatted:

  * apps/ewallet_db/lib/ewallet_db/validator.ex
  * apps/ewallet/lib/ewallet/gates/transaction_request_gate.ex
  * mix.exs
  * apps/ewallet/lib/ewallet/web/paginator.ex
  * apps/admin_api/lib/admin_api/v1/controllers/reset_password_controller.ex
  * apps/ewallet_db/lib/ewallet_db/account.ex
  * apps/local_ledger_db/lib/local_ledger_db/minted_token.ex
```

We're in the middle of auditing Jenkins security for opening to public, sorry for the inconvenient!

**Comment by unnawut** [comment]
It's still failing somehow...

```
** (Mix) mix format failed due to --check-formatted.
The following files were not formatted:

  * apps/ewallet/test/ewallet/gates/transaction_consumption_consumer_gate_test.exs
  * apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_controller_test.exs
```

**Comment by InoMurko** [comment]
hm hm @unnawut my `mix format `doesn't detect `transaction_controller_test.exs` problems.
`IEx 1.6.5 (compiled with OTP 19)`

**Comment by unnawut** [comment]
@sirn Looks like it's caused by version difference indeed. Elixir is [updating its `mix format` rules quite frequently](https://github.com/elixir-lang/elixir/releases).

Can we simply bump up the Elixir version in https://github.com/omisego-images/docker-elixir/blob/master/1.6-otp20-stretch/Dockerfile ? Not sure if that might cause incompatibility issues anywhere else.

**Comment by sirn** [comment]
@unnawut will do

**Comment by unnawut** [comment]
You're right. The key (2nd parameter) passed into `Changeset.add_error/4` should be an atom.

I believe we could easily use `Enum.split(attrs, 1)` or similar to pass an affected field as the key (2nd parameter) and pass the rest as the message (3rd parameter) to `Changeset.add_error/4` instead.

File: apps/ewallet/lib/ewallet/seeders/cli_seeder.ex [hunk 1]
```diff
@@ -43,8 +43,8 @@ defmodule EWallet.Seeder.CLI do
       |> Seeder.argsline_for()
       |> process_argsline()
 
-    IO.puts("\n-----\n")
-    IO.gets(@confirm_message)
+    _ = IO.puts("\n-----\n")
```
Comment (T-Dnzt) [inline apps/ewallet/lib/ewallet/seeders/cli_seeder.ex]: Interesting! Does each function call should be patterned matched in this way?

Comment (InoMurko) [reply apps/ewallet/lib/ewallet/seeders/cli_seeder.ex]: only when you wan't to ignore the returned value

File: apps/ewallet_db/lib/ewallet_db/account.ex [hunk 2]
```diff
@@ -108,7 +108,8 @@ defmodule EWalletDB.Account do
     end
   end
 
-  @spec avatar_changeset(changeset :: Ecto.Changeset.t(), attrs :: map()) :: Ecto.Changeset.t()
+  @spec avatar_changeset(changeset :: Ecto.Changeset.t() | %Account{}, attrs :: map()) ::
+          Ecto.Changeset.t() | no_return() | none()
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/account.ex]: I'm not too familiar with typespec, but isn't `no_return()` and `none()` the same? ([ref](https://hexdocs.pm/elixir/typespecs.html#built-in-types))

Also applies to the remainder of the PR.

File: apps/ewallet_db/lib/ewallet_db/helpers/crypto.ex [hunk 3]
```diff
@@ -4,7 +4,7 @@ defmodule EWalletDB.Helpers.Crypto do
   """
   alias Plug.Crypto
 
-  @spec generate_key(integer) :: binary
+  @spec generate_key(Integer.t()) :: binary()
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/helpers/crypto.ex]: Would `integer()` be more accurate than `Integer.t()`?

I also couldn't find a type defined on [`Integer`](https://github.com/elixir-lang/elixir/blob/v1.6.5/lib/elixir/lib/integer.ex).

File: apps/ewallet_db/lib/ewallet_db/token.ex [hunk 4]
```diff
@@ -154,7 +154,7 @@ defmodule EWalletDB.Token do
   @doc """
   Retrieves a token using one or more fields.
   """
-  @spec get_by(fields :: map(), opts :: keyword()) :: %Token{} | nil
+  @spec get_by(fields :: keyword(), opts :: keyword()) :: %Token{} | nil
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/token.ex]: Actually this supports both `keyword()` and `map()` as the 1st argument.

File: apps/ewallet/lib/ewallet/web/email_validator.ex [hunk 5]
```diff
@@ -7,7 +7,7 @@ defmodule EWallet.EmailValidator do
   @doc """
   Checks whether the email address looks correct.
   """
-  @spec valid?(String.t()) :: boolean()
+  @spec valid?(String.t()) :: String.t() | boolean()
```
[the remark this answers is not in the document]
Comment (InoMurko) [reply apps/ewallet/lib/ewallet/web/email_validator.ex]: correct

File: apps/admin_api/lib/admin_api/v1/controllers/user_controller.ex [hunk 6]
```diff
@@ -60,5 +60,5 @@ defmodule AdminAPI.V1.UserController do
   end
 
   # Responds when the user is not found
-  defp respond_single(nil, conn), do: handle_error(conn, :user_id_not_found)
+  defp respond_single(_, conn), do: handle_error(conn, :user_id_not_found)
```
[the remark this answers is not in the document]
Comment (InoMurko) [reply apps/admin_api/lib/admin_api/v1/controllers/user_controller.ex]: correct but also below 
`defp respond_single({:error, changeset}, conn) do` can never match.

File: apps/ewallet/lib/ewallet/web/preloader.ex [hunk 7]
```diff
@@ -4,11 +4,12 @@ defmodule EWallet.Web.Preloader do
   It takes in a list of associations to preload as a list of atoms.
   """
   alias EWalletDB.Repo
+  alias EWalletDB.Transfer
 
   @doc """
   Preload the given list of associations.
   """
-  @spec to_query(Ecto.Query.t(), List.t()) :: {Ecto.Query.t()}
+  @spec to_query(Ecto.Query.t() | Transfer.t(), List.t()) :: {Ecto.Query.t()}
```
[the remark this answers is not in the document]
Comment (InoMurko) [reply apps/ewallet/lib/ewallet/web/preloader.ex]: correct

**Comment by InoMurko** [comment]
Yeah, not many left! :)

**Pull request merged**